### PR TITLE
bitcoin-cli: Provide a better error message when bitcoind is not running

### PR DESCRIPTION
**Pull request by practicalswift**
Provide a better `bitcoin-cli` error message when `bitcoind` is not running.

Before this patch:

```
$ killall -9 bitcoind
$ bitcoin-cli -testnet echo 'hello world'
error: Could not locate RPC credentials. No authentication cookie could be found, and RPC password is not set.  See -rpcpassword and -stdinrpcpass.  Configuration file: (/root/.bitcoin/bitcoin.conf)
```

After this patch:

```
$ killall -9 bitcoind
$ bitcoin-cli -testnet echo 'hello world'
error: Could not connect to the server 127.0.0.1:18332

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.
```

### CODE REVIEW

**Review by randolf — APPROVED** [review]

**Comment by jonasschnelli** [comment]
I guess you need to fix the `interface_bitcoin_cli.py` test.

**Comment by practicalswift** [comment]
@jonasschnelli Oh, thanks! Fixed. Please re-review :-)

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/12652/commits/de04fde534997f15235ded46912806d47c615c23

**Comment by cdecker** [comment]
utACK de04fde

**Comment by laanwj** [comment]
I remember making the code like this to avoid certain edge cases around the authentication credentials and connectability.

Can we please add tests for this? This logic was already complicated, and this makes it even more complicated.

File: src/bitcoin-cli.cpp [hunk 1]
```diff
@@ -358,11 +356,21 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
...
+    if (response.status == 0) {
+        std::string responseErrorMessage;
+        if (response.error != -1) {
+            responseErrorMessage = strprintf(" (error code %d - \"%s\")", response.error, http_errorstring(response.error));
+        }
+        throw CConnectionFailed(strprintf("Could not connect to the server %s:%d%s\n\nMake sure the bitcoind server is running and that you are connecting to the correct RPC port.", host, port, responseErrorMessage));
+    } else if (response.status == HTTP_UNAUTHORIZED) {
+        if (failedToGetAuthCookie) {
+            throw std::runtime_error(strprintf(
+                _("Could not locate RPC credentials. No authentication cookie could be found, and RPC password is not set.  See -rpcpassword and -stdinrpcpass.  Configuration file: (%s)"),
```
Comment (MarcoFalke) [inline src/bitcoin-cli.cpp]: Mind to add a test for this?

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -358,11 +356,21 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
...
-        throw CConnectionFailed(strprintf("couldn't connect to server: %s (code %d)\n(make sure server is running and you are connecting to the correct RPC port)", http_errorstring(response.error), response.error));
-    else if (response.status == HTTP_UNAUTHORIZED)
-        throw std::runtime_error("incorrect rpcuser or rpcpassword (authorization failed)");
-    else if (response.status >= 400 && response.status != HTTP_BAD_REQUEST && response.status != HTTP_NOT_FOUND && response.status != HTTP_INTERNAL_SERVER_ERROR)
+    if (response.status == 0) {
+        std::string responseErrorMessage;
+        if (response.error != -1) {
+            responseErrorMessage = strprintf(" (error code %d - \"%s\")", response.error, http_errorstring(response.error));
+        }
+        throw CConnectionFailed(strprintf("Could not connect to the server %s:%d%s\n\nMake sure the bitcoind server is running and that you are connecting to the correct RPC port.", host, port, responseErrorMessage));
```
Comment (MarcoFalke) [inline src/bitcoin-cli.cpp]: Same here. Should come with a test.

**Comment by practicalswift** [comment]
Tests added please re-review :-)

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (practicalswift) [reply src/bitcoin-cli.cpp]: Added :-)

File: src/bitcoin-cli.cpp (hunk 2 above)
Comment (practicalswift) [reply src/bitcoin-cli.cpp]: Added :-)

**Comment by laanwj** [comment]
utACK 8b2ef27ff9be20fc1a53d8804b403859251f1fca / tests look good to me, thanks for adding them.

**Pull request merged**